### ISSUE TITLE
Always use Named Arguments with booleans

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#3534]
## Expected Behavior of the rule

Noncompliant
```kotlin
User("Brais", false)

getUser(id, true)
```

Compliant
```kotlin
User("Brais", isActive = false)

getUser(id, forceRefresh = true)

getUser(id, refresh)
```

## Context
A boolean as a parameter creates difficult code to review:

For example if you have this `data class`:
```kotlin
User(
  val name: String,
  val isActive: Boolean,
  val isPro: Boolean,
)
```

This code is really dificult to understand: `User("Brais", false, true)`. It's kind of easy to understand that `"Brais"` is the name of the user. But it's impossible to know if this user `isActive` or if it `isPro`. If you want to know it you need check the User definition.

With only one parameter the code is a bit better but the problem is still there.

I know that this is a bit subjective and there are other good cases to use argument parameters not adressed here. But I think that the boolean is a real common case that a lot of detekt users can use.

**Comment by chao2zhang** [comment]
Meanwhile, there are good discussions on https://youtrack.jetbrains.com/issue/KT-14934 and I feel like there are a variety of needs for enforcing named parameter usage. I can see why Jetbrains don't want to step in but we could collect more user feedbacks.

**Comment by cortinico** [comment]
> I know that this is a bit subjective and there are other good cases to use argument parameters not adressed here. But I think that the boolean is a real common case that a lot of detekt users can use.

Is this going to be addressing only data classes?
As I can imagine several scenarios where named parameters are not desired:

```
.setWarningAsError(true)
```

vs

```
.setWarningAsError(warningAsError = true)
// or 
.setWarningAsError(enabled = true)
```

I believe the first snipped should be preferred. Plus by enforcing the naming parameters, users would have to explicitly use parameter names that are not always nicely named (as historically they were not part of the method signatures).

**Comment by BraisGabin** [comment]
> We can also build a new rule to enforce named arguments for all adjacent parameters with the same type.

I really like this idea! But, only adjacent? Why is `foo("hi", true, "chao")` better than `foo("hi",  "chao", true)`? Couldn't we just enforce named arguments if your function have two parameters of the same type?

This rule will also promote the usage of `inline class`es as types as described here: https://jakewharton.com/inline-classes-make-great-database-ids/

Down side: I can see cases where this is not desired: `fun plus(a: Int, b: Int)` I don't care the order of the parameters here...

> Is this going to be addressing only data classes?
> As I can imagine several scenarios where named parameters are not desired:
> 
> ```
> .setWarningAsError(true)
> ```

No, I was not talking just about data classes. But your sample is really good. I agree that in your case the named argument is not desired. 🤔 maybe we should add a configuration (`on` by default) that ignores the cases where the only argument is a boolean?

**Comment by chao2zhang** [comment]
Adjacent parameters with the same type are based on a real-life example. We did have Java code that swaps the position of two Strings and caught them much later.

I guess theoretically `foo("hi", true, "chao")` is possible to be confused with `foo("chao", true, "hi")`, but practically `foo("hi", "chao", true)` is more likely to be confused with `foo("chao", "hi", true)`.

**Comment by BraisGabin** [comment]
Agree, I think that if we end up implementing this rule we could add a flag to enable or disable the "adjacent" part. It have sense.

**Comment by VovaStelmashchuk** [comment]
@BraisGabin I try to implement the rule for the issue, can you review ?

**Comment by stunstunstun** [comment]
@BraisGabin @VovaStelmashchuk  Is there a possibility that this issue will resume?

**Comment by cortinico** [comment]
This is up for grab :+1:

**Comment by VovaStelmashchuk** [comment]
@stunstunstun currently I am not working on the issue. You can try to implement solution on your side, also you can rely on my help with code review or help in implementation the soluction.

**Comment by atulgpt** [comment]
Hi @BraisGabin / @VovaStelmashchuk / @cortinico I have created PR at https://github.com/detekt/detekt/pull/6055
